packages/react-reconciler/src/ReactFiberBeginWork.new.js

```diff
@@ -2157,44 +1955,276 @@ function updateSuspenseComponent(
         // Skip the primary children, and continue working on the
         // fallback children.
         workInProgress.memoizedState = mountSuspenseState(renderExpirationTime);
-        workInProgress.child = primaryChildFragment;
         return fallbackChildFragment;
       } else {
         // Still haven't timed out. Continue rendering the children, like we
         // normally do.
-        workInProgress.memoizedState = null;
         const nextPrimaryChildren = nextProps.children;
-        return (workInProgress.child = reconcileChildFibers(
+        const primaryChildFragment = updateSuspensePrimaryChildren(
+          current,
           workInProgress,
-          currentPrimaryChild,
           nextPrimaryChildren,
           renderExpirationTime,
-        ));
+        );
+        workInProgress.memoizedState = null;
+        return primaryChildFragment;
       }
     }
   }
 }
 
+function mountSuspensePrimaryChildren(
+  workInProgress,
+  primaryChildren,
+  renderExpirationTime,
+) {
+  const mode = workInProgress.mode;
+  const primaryChildFragment = createFiberFromFragment(
+    primaryChildren,
+    mode,
+    renderExpirationTime,
+    null,
+  );
+  primaryChildFragment.return = workInProgress;
+  workInProgress.child = primaryChildFragment;
+  return primaryChildFragment;
+}
+
+function mountSuspenseFallbackChildren(
+  workInProgress,
+  fallbackChildren,
+  renderExpirationTime,
+) {
+  const mode = workInProgress.mode;
+
+  const progressedPrimaryFragment: Fiber | null = workInProgress.child;
+
+  let primaryChildFragment;
+  let fallbackChildFragment;
+  if ((mode & BlockingMode) === NoMode && progressedPrimaryFragment !== null) {
+    // In legacy mode, we commit the primary tree as if it successfully
+    // completed, even though it's in an inconsistent state.
+    primaryChildFragment = progressedPrimaryFragment;
+    primaryChildFragment.childExpirationTime_opaque = NoWork;
+
+    if (enableProfilerTimer && workInProgress.mode & ProfileMode) {
+      // Reset the durations from the first pass so they aren't included in the
+      // final amounts. This seems counterintuitive, since we're intentionally
+      // not measuring part of the render phase, but this makes it match what we
+      // do in Concurrent Mode.
+      primaryChildFragment.actualDuration = 0;
+      primaryChildFragment.actualStartTime = -1;
+      primaryChildFragment.selfBaseDuration = 0;
+      primaryChildFragment.treeBaseDuration = 0;
+    }
+
+    // Transfer the effects from the primary tree
```

Comment:
Why do we have to transfer the effects now but didn't before? I'd expect them to be transferred automatically after we complete, since the node is a child.

Reply:
The comment isn't quite right. It's to remove the current fallback fiber from the deletion list, which was added in the first pass. However, that's only relevant to updates, and this is a mount path, so you're right that I can remove this here. But it is necessary in the `updateSuspenseFallbackChildren` path.

What we used to do is only add the fallback fiber to the effect list once we get to the complete phase: https://github.com/facebook/react/pull/18711/files/d4ac75f601ce71d73893e895892fa6a3d1dbbaf4..cd7e3948d3a1159ec06ae1a326c0ea2547656c6c#diff-c29bccbec7fe889841f1b254f08d97c7L891-L910

I opted for doing it here instead since this means we don't have to do anything special in Concurrent Mode. It's only weird in legacy mode because we don't create a fresh work-in-progress fiber for the second pass, which would reset the effect list. Similarly, we don't reuse effects from the first pass.

Another approach which I just thought of would be to clear the Deletion bit from the tag but keep it in the list. Slightly less code but slightly more computation later.

I'll delete this from the mount path, leave the update path as-is, and fix the comment.